### PR TITLE
*: support recover relay log file with a subset of metadata

### LINKED ISSUE #443: fial to truncate GTIDs if relay.meta out of date before killed when recovering
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

    1. `kill -9` for DM-worker.
    2. restart DM-worker

2. What did you expect to see?

    - restart DM-worker successfully.

3. What did you see instead?

    - image

4. Versions of the cluster

    - DM v1.0.2

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

fix #443 , this should be release under v1.0.4.

### What is changed and how it works?

- support recover relay log file with a subset of metadata
    - check whether the metadata is a subset of the relay log file

- also update `go-mysql` to fix `GTIDSet.Equal`.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change
 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @lichunzhu PTAL

File: pkg/binlog/event/common.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func GenCommonFileHeader(flavor string, serverID uint32, gSet gtid.Set) ([]*repl
 			Flags:     defaultHeaderFlags,
 		}
 		latestPos   = uint32(len(replication.BinLogFileHeader))
-		prevGTIDsEv *replication.BinlogEvent // for MySQL, this will be a GenericEvent
+		prevGTIDsEv *replication.BinlogEvent
```
Comment (WangXiangUSTC) [inline pkg/binlog/event/common.go]: is the new go-mysql change the rule？

Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: Yes, new go-mysql adds a new `PreviousGTIDsEvent ` struct to replace `GenericEvent with EventType=PREVIOUS_GTIDS_EVENT`.

**Comment by csuzhangxc** [comment]
> do we need to add kill -9 dm-worker in integration test?

@WangXiangUSTC like a special type of chaos test? In fact, this is a litter hard to simulate it by simply kill the process (when just writing the data).

**Comment by WangXiangUSTC** [comment]
> > do we need to add kill -9 dm-worker in integration test?
> 
> @WangXiangUSTC like a special type of chaos test? In fact, this is a litter hard to simulate it by simply kill the process (when just writing the data).

ok , we can ignore it now

**Comment by WangXiangUSTC** [comment]
LGTM

File: pkg/terror/error_list.go [hunk 2]
```diff
@@ -642,7 +642,7 @@ var (
 	ErrBinlogWriterWriteDataLen      = New(codeBinlogWriterWriteDataLen, ClassFunctional, ScopeInternal, LevelHigh, "data length %d")
 	ErrBinlogWriterFileNotOpened     = New(codeBinlogWriterFileNotOpened, ClassFunctional, ScopeInternal, LevelHigh, "file %s not opened")
 	ErrBinlogWriterFileSync          = New(codeBinlogWriterFileSync, ClassFunctional, ScopeInternal, LevelHigh, "sync file")
-	ErrBinlogPrevGTIDEvNotValid      = New(codeBinlogPrevGTIDEvNotValid, ClassFunctional, ScopeInternal, LevelHigh, "PreviousGTIDsEvent should be a GenericEvent in go-mysql, but got %T")
+	ErrBinlogPrevGTIDEvNotValid      = New(codeBinlogPrevGTIDEvNotValid, ClassFunctional, ScopeInternal, LevelHigh, "PreviousGTIDsEvent should be a PreviousGTIDsEvent in go-mysql, but got %T")
```
Comment (lichunzhu) [inline pkg/terror/error_list.go]: I can't understand this error message.

Comment (csuzhangxc) [reply pkg/terror/error_list.go]: updated in https://github.com/pingcap/dm/pull/530/commits/08679ddf4ead683a6a343c94c837a29623aff147.

**Pull request merged**